File: Algorithms/C++/Medium/106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if (inorder.empty() || postorder.empty() || inorder.size() != postorder.size()) {
            return NULL;
        }

        return helper(inorder, 0, inorder.size() - 1, postorder, 0, postorder.size() - 1);
    }
private:
    TreeNode* helper(vector<int>& inorder, int inorder_start, int inorder_end, vector<int>& postorder, int postorder_start, int postorder_end) {
        if (inorder_start > inorder_end) {
            return NULL;
        }

        int root_val = postorder[postorder_end];
        TreeNode* root = new TreeNode(root_val);

        int inorder_root_index;
        for (inorder_root_index = inorder_start; inorder_root_index <= inorder_end && inorder[inorder_root_index] != root_val; inorder_root_index++) {
            ;
        }
        int len = inorder_root_index - inorder_start;

        root->left = helper(inorder, inorder_start, inorder_root_index - 1, postorder, postorder_start, postorder_start + len - 1);
        root->right = helper(inorder, inorder_root_index + 1, inorder_end, postorder, postorder_start + len - 1 + 1, postorder_end - 1);

        return root;
    }
};

File: Algorithms/Swift/Medium/106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func buildTree(inorder: [Int], _ postorder: [Int]) -> TreeNode? {
        if inorder.isEmpty || postorder.isEmpty || inorder.count != postorder.count {
            return nil
        }
        
        return helper(inorder: inorder, inorderStartIndex: 0, inorderEndIndex: inorder.count - 1, postorder: postorder, postorderStartIndex: 0, postorderEndIndex: postorder.count - 1)
    }
    
    private func helper(inorder inorder: [Int], inorderStartIndex: Int, inorderEndIndex: Int, postorder: [Int], postorderStartIndex: Int, postorderEndIndex: Int) -> TreeNode? {
        if inorderStartIndex > inorderEndIndex {
            return nil
        }
        
        let rootVal = postorder[postorderEndIndex]
        let root    = TreeNode(rootVal)
        
        var inorderIndex = 0
        for i in inorderStartIndex...inorderEndIndex {
            if inorder[i] == rootVal {
                inorderIndex = i
                break
            }
        }
        
        let length = inorderIndex - inorderStartIndex
        
        root.left  = helper(inorder: inorder, inorderStartIndex: inorderStartIndex, inorderEndIndex: inorderIndex - 1, postorder: postorder, postorderStartIndex: postorderStartIndex, postorderEndIndex: postorderStartIndex + length - 1) 
        root.right = helper(inorder: inorder, inorderStartIndex: inorderIndex + 1, inorderEndIndex: inorderEndIndex, postorder: postorder, postorderStartIndex: postorderStartIndex + length, postorderEndIndex: postorderEndIndex - 1)
        
        return root
    }
}
